rbd-fuse: look for ceph.conf in standard locations

rishabh-d-dave opened the pull request:
When ceph.conf is absent in /etc/ceph/ and it is not being passed as an
argument, look at the environment variable $CEPH_CONF, and in ~/.ceph/,
current working directory to find one. If all these are present, unlike
now, do report it to the user.

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave commented:
Though not sure if this is the most correct way to do it.

dillaman commented:
@rishabh-d-dave See [1] for an example of a client RBD-based daemon processing command-line arguments/environment variables for initializing Ceph

[1] https://github.com/ceph/ceph/blob/master/src/tools/rbd_nbd/rbd-nbd.cc#L1017

trociny commented:
@rishabh-d-dave rbd-nbd.cc is probably a little complicated example for this case. Actually, it looks to me it could be fixed here in `connect_to_cluster()` function. Currently it has:
```
rados_conf_parse_env(*pcluster, NULL);
r = rados_conf_read_file(*pcluster, rbd_options.ceph_config);
```
Which means that it properly parse environment (looking for options in  `CEPH_ARGS` variable), but does not parse command line options (e.g. looking for the config file name), and then calls `rados_conf_read_file` with the config file explicitly set, i.e. it does not use built-in facility to find the config file in common locations.

I suppose if we provided `argc` and `argv` somehow to `connect_to_cluster()` function (e.g. via `rbd_options` global structure) and changed it to something like this:
```
r = rados_conf_parse_env(*pcluster, NULL);
r = rados_conf_parse_argv(*pcluster, argc, argv);
r = rados_conf_read_file(*pcluster, NULL);  // Note, NULL instead of the explicitly set path
```
it could do the trick.

I have not tried this so not sure if it will work as I expect but could you try something like this?

rishabh-d-dave commented:
@trociny It didn't work for me. Error I see is same as the usual one -

`$ sudo ls /rbd_images/`
`ls: cannot access '/rbd_images/': Transport endpoint is not connected`

I've attached the changes I made (as .txt file, .patch didn't work).

How can I take look what went wrong with callbacks? Using gdb or printf()s/fprintf()s didn't work me.


[pass-argv-to-func-connect-to-cluster.txt](https://github.com/ceph/ceph/files/1766612/pass-argv-to-func-connect-to-cluster.txt)

trociny commented:
> pass-argv-to-func-connect-to-cluster.txt

In the patch I don't see that `rbd_options.argc` and `rbd_options.argv` are set to those that come to `main()`.

Also, you don't check the returned (`r`) code.

And why not push the changes to this PR? It is much more convenient for review. You can always make a copy of your original branch in case you want to revert back.

rishabh-d-dave commented:
@trociny Sorry about last time; made a mistake while uploading.

trociny commented:
Ok. Playing with this, `rados_conf_parse_argv` does not look useful and we still need to provide `rbd_options.ceph_config` to `rados_conf_read_file` when it is specified.

It looks like one of the problems actually is that that `rbd_options.ceph_config` is initialized to '/etc/ceph/ceph.conf' instead of NULL. So when it is not specified via command line, '/etc/ceph/ceph.conf' will be always used. So just fixing `rbd_options` initialization
```
struct rbd_options rbd_options = {NULL, (char *)"rbd", NULL};
```
makes things much better: all possibilities (`$CEPH_CONF`, ceph.conf in the current directory and in /etc/ceph) are used. There is still a problem with ceph.conf in the current directory: when the rbd-fuse is started in foreground (-f or -d option) it properly looks for the file in the local directory. But when it is started in background it changes the working directory to '/' so it looks for ceph.conf file here. (You can check this creating a symlink to ceph.conf in / to see that it works after this).

To fix this problem it looks like you need to move `rados_create`, `rados_conf_parse_env`, `rados_conf_read_file` from `connect_to_cluster` to `main` (after `fuse_opt_parse`, use `cluster` global instead of `pcluster`)

@rishabh-d-dave Are you willing to try this?

rishabh-d-dave commented:
@trociny Yeah, I'll try it.

rishabh-d-dave commented:
It worked fine for me but it couldn't read ~/.ceph/ceph.conf unless it was explicitly passed -

```
$ export | grep CEPH_CONF
$ ls ./ceph.*
./ceph.conf.bak
$ ls /etc/ceph/
rbdmap
$ ls ~/.ceph/
ceph.conf
$ sudo ./bin/rbd-fuse ~/rbd_images
Error reading Ceph config file: No such file or directory
2018-03-04 17:35:25.679 7f9b0f93e280 -1 Errors while parsing config file!
2018-03-04 17:35:25.679 7f9b0f93e280 -1 parse_file: cannot open /etc/ceph/ceph.conf: (2) No such file or directory
2018-03-04 17:35:25.679 7f9b0f93e280 -1 parse_file: cannot open ~/.ceph/ceph.conf: (2) No such file or directory
2018-03-04 17:35:25.679 7f9b0f93e280 -1 parse_file: cannot open ceph.conf: (2) No such file or directory
$ 
[rishabh@x220 build]$ sudo ./bin/rbd-fuse ~/rbd_images -c ~/.ceph/ceph.conf  #this works
```

I also have ceph.conf in /root/.ceph/ and tried to run it as the root user but I get the same behaviour. It could not open the file though (in the following snippet gdb is at src/dmclock/sim/src/ConfUtils.cc) -
```
93	  FILE *fp = fopen(fname.c_str(), "r");
(gdb) p fname
$5 = "~/.ceph/ceph.conf"
(gdb) n
94	  if (!fp) {
(gdb) n
95	    ostringstream oss;
```

fopen() doesn't know that the tilde has a special meaning. I will write a (separate?) patch for that too.

trociny commented:
> [rishabh@x220 build]$ sudo ./bin/rbd-fuse ~/rbd_images -c ~/.ceph/ceph.conf  #this works

The tilde is expanded by the shell, before the command is executed, so your home directory is passed to sudo  command.

> 2018-03-04 17:35:25.679 7f9b0f93e280 -1 parse_file: cannot open ~/.ceph/ceph.conf: (2) No such file or directory

This looks like a bug, not related to rbd-fuse. For sure it needs to be expanded before trying to open it. I think `md_config_t::parse_config_files` should do this similarly how it does with "$data_dir". So it was broken or never worked. Probably you are the first who tried this. Could you please report this to the tracker?

trociny commented on src/rbd_fuse/rbd-fuse.cc:
```diff
@@ -879,11 +867,36 @@ connect_to_cluster(rados_t *pcluster)
+		return -1;
+	}
+
+	r = rados_conf_parse_argv(cluster, argc, (const char **) argv);
```
I don't think it is useful as is.

It could be useful, because it would allow to pass common ceph configuration options, like `--id`, ` --cluster`, similarly to what other ceph tools, like `rbd` or `rbd-nbd`, do. The problem here though that earlier `argv` is passed to `fuse_opt_parse`, and I suppose it will fail when it see an unknown (to it) option.

So I see two ways:

1) (simpler) Just don't use `rados_conf_parse_argv` here as it is not useful -- your patch will fix the problem with the config, and this is already an improvement.
2) (more correct in my opinion) I see rados API additionally provides this function:
```
// like above, but return the remainder of argv to contain remaining
// unparsed args.  Must be allocated to at least argc by caller.
// remargv will contain n <= argc pointers to original argv[], the end
// of which may be NULL

extern "C" int rados_conf_parse_argv_remainder(rados_t cluster, int argc,
                                               const char **argv,
                                               const char **remargv)
```
It looks like it is for the cases like this. If we used  `rados_conf_parse_argv_remainder` instead of `rados_conf_parse_argv` and did this before `FUSE_ARGS_INIT`, and then passed `remargv` to `FUSE_ARGS_INIT` instead of `argv` then I suppose it would do the trick.

@rishabh-d-dave  do you want to try (2)? Or you can just provide (1) leaving (2) to be implemented later.

rishabh-d-dave replied:
I will try (2), since it looks more proper.   :)

And will incorporate the rest of improvements you pointed out.

rishabh-d-dave commented:
I had to improvise a bit -- `rados_conf_read_file()` can only be called only after `fuse_opt_parse()` sets `rbd_options.ceph_config` to the path passed with "-c". So I had to move it `main()` just before calling `fuse_main()`.

BTW, should I merge this patch [1] with current one? IMO, they are related but can't be treated as a single change.

[1] https://github.com/ceph/ceph/pull/20614#issuecomment-369091676

rishabh-d-dave commented:
@trociny I have diverted a bit from your instructions -- `pre_init_cluster()` has been named `init_cluster()`, `init_cluster()` and `post_init_cluster()` return remargc and nothing respectively (this allows to call `exit(1)` inside these functions so that number of lines are reduced) and `rados_conf_parse_argv()` has not been called in `post_init_cluster()` as it wasn't required.

trociny commented on src/rbd_fuse/rbd-fuse.cc:
```diff
@@ -877,13 +865,59 @@ connect_to_cluster(rados_t *pcluster)
+		exit(1);
+	}
+
+	for (remargc = 0; remargv[remargc] != NULL and remargc < argc;
```
I think you need to check for `remargc < argc` before trying `remargv[remargc]`, i.e. swap these predicated, otherwise you are reading out of the buffer when `remargc == argc`.

trociny commented on src/rbd_fuse/rbd-fuse.cc:
```diff
@@ -877,13 +865,59 @@ connect_to_cluster(rados_t *pcluster)
+		simple_err("Error reading Ceph config file", r);
+		exit(1);
+	}
+
```
I wanted additional `rados_conf_parse_argv` after `rados_conf_read_file` to make command line options override config options (which is expected behavior).

Assuming you have "admin socket" and "log file" options specified in ceph.conf, try this:
```
./bin/rbd-fuse --log-file=/tmp/rbd-fuse.log --admin-socket=/tmp/rbd-fuse.asok  /mnt
```
Your current version will still use the config file options.

trociny commented on src/rbd_fuse/rbd-fuse.cc:
```diff
@@ -877,13 +865,59 @@ connect_to_cluster(rados_t *pcluster)
 {
-	struct fuse_args args = FUSE_ARGS_INIT(argc, argv);
+	char *remargv[argc] = {NULL};
+	int remargc = init_cluster(argc, argv, remargv);
```
nit: I would still prefer my version. When a function returns something, I always suppose this is the most important thing, e.g. the result or status (error code) of its work. And if a function may fail I would like to see this just looking on how it is used, without need to read this function code.

Saving one line or two is not always good if it makes the code less readable. But taking that this is rather a small  code, it is not important here, so it is up to you what signature to use.

trociny commented:
And I think you need to add 'Fixes: http://tracker.ceph.com/issues/12219' to the log message.

trociny commented:
> BTW, should I merge this patch [1] with current one? IMO, they are related but can't be treated as a single change.
> [1] #20614 (comment)

If `rados_conf_read_file` does not work correctly with relative paths, in my opinion it should be fixed, so all its users win (if I don't miss a reason why it is this way).  Anyway I think this is a separate problem, and having a separate PR for it looks right to me.

rishabh-d-dave commented:
> If rados_conf_read_file does not work correctly with relative paths, in my opinion it should be fixed, so all its users win (if I don't miss a reason why it is this way). Anyway I think this is a separate problem, and having a separate PR for it looks right to me.

Well, that behaviour gets automatically fixed with this patch. Initially, when these `rados_conf_read_file()` was being called by libfuse,  CWD would change to `/`. So, passing relative paths wouldn't make sense. But now since it is being called by main, instead of libfuse, CWD stays the same and relative paths make sense to the `rados_conf_read_file()`. In short, that bug, too, got fixed with this one.

@trociny Should this be mentioned in commit message or elsewhere, since it also fixes this issue - https://tracker.ceph.com/issues/23134?

trociny commented:
> Should this be mentioned in commit message or elsewhere, since it also fixes this issue - https://tracker.ceph.com/issues/23134?

Jason set 23134 as duplicate of 12219. So just 'Fixes: http://tracker.ceph.com/issues/12219' is enough but I don't mind if you mention this in the commit message.

rishabh-d-dave commented:
@trociny I'll surely investigate the issue. But I'm not sure if I can figure out before you do.  :)

rishabh-d-dave commented:
@trociny I've tested with ceph.conf in different locations as well as with absolute and relative paths as an argument to -c and I don't see any issues. It seems to be working fine for me except when ceph.conf is at ~/.ceph. Are you talking about that?

trociny commented:
No, I am talking about this:
```
maha:~/ceph/ceph/build% sudo ./bin/rbd-fuse /mnt
maha:~/ceph/ceph/build% sudo ls /mnt
test  test1  test2  test3  test4  test5  test6  test7
maha:~/ceph/ceph/build% sudo umount /mnt        
maha:~/ceph/ceph/build% dmesg| tail -1          
[3383004.447127] rbd-fuse[18921]: segfault at 7f8ce7c219d0 ip 00007f8ceaa7f8cc sp 00007fff12db8f10 error 4 in libpthread-2.27.so[7f8ceaa77000+19000]
```
So it runs ok, but when the process is terminated (on unmount) it segfaults. It might not see a coredump because it is deamonized. But you can get the backtrace this way. Start rbd-fuse, attach to it with `gdb -p pid`, issue `continue` command in the gdb, unmount rbd-fuse directory. After this you should see the crush in the gdb:
```Thread 1 "rbd-fuse" received signal SIGSEGV, Segmentation fault.
0x00007f57039e68cc in __pthread_timedjoin_ex () from /lib64/libpthread.so.0
(gdb) bt
#0  0x00007f57039e68cc in __pthread_timedjoin_ex () from /lib64/libpthread.so.0
#1  0x00007f570408564c in Thread::join (this=0x55ad8b2685e0, prval=prval@entry=0x0)
    at /home/mgolub/ceph/ceph/src/common/Thread.cc:164
#2  0x00007f5703ea8fd3 in ceph::logging::Log::stop (this=<optimized out>)
    at /home/mgolub/ceph/ceph/src/log/Log.cc:449
#3  0x00007f570407da18 in CephContext::~CephContext (this=0x55ad8b1f9920, __in_chrg=<optimized out>)
    at /home/mgolub/ceph/ceph/src/common/ceph_context.cc:671
#4  0x00007f570407dcd4 in CephContext::put (this=0x55ad8b1f9920)
    at /home/mgolub/ceph/ceph/src/common/ceph_context.cc:690
#5  0x00007f570d82ca74 in std::function<void (CephContext*)>::operator()(CephContext*) const (
    __args#0=<optimized out>, this=<optimized out>) at /usr/include/c++/7/bits/std_function.h:706
#6  std::unique_ptr<CephContext, std::function<void (CephContext*)> >::~unique_ptr() (
    this=0x55ad8b239ed0, __in_chrg=<optimized out>) at /usr/include/c++/7/bits/unique_ptr.h:268
#7  librados::RadosClient::~RadosClient (this=0x55ad8b239ec0, __in_chrg=<optimized out>)
    at /home/mgolub/ceph/ceph/src/librados/RadosClient.cc:467
#8  0x00007f570d82cac9 in librados::RadosClient::~RadosClient (this=0x55ad8b239ec0, 
    __in_chrg=<optimized out>) at /home/mgolub/ceph/ceph/src/librados/RadosClient.cc:474
#9  0x00007f570d7d77ad in rados_shutdown (cluster=0x55ad8b239ec0)
    at /home/mgolub/ceph/ceph/src/librados/librados.cc:2885
#10 0x00007f570d548159 in fuse_fs_destroy () from /lib64/libfuse.so.2
#11 0x00007f570d5481ac in ?? () from /lib64/libfuse.so.2
#12 0x00007f570d54a4e2 in ?? () from /lib64/libfuse.so.2
#13 0x00007f570d550603 in fuse_session_destroy () from /lib64/libfuse.so.2
#14 0x00007f570d549506 in fuse_destroy () from /lib64/libfuse.so.2
#15 0x00007f570d551a40 in ?? () from /lib64/libfuse.so.2
#16 0x00007f570d55225b in ?? () from /lib64/libfuse.so.2
#17 0x000055ad8a6fe602 in main (argc=2, argv=0x7ffff5cbf488)
    at /home/mgolub/ceph/ceph/src/rbd_fuse/rbd-fuse.cc:935
```

trociny commented:
There is another issue with our approach: when filtering out ceph common options, it also filters out '-f' and '-d'  options, because they are recognized as ceph options, and they are not passed to the fuse option parser. As a result now '-f' and '-d' do not work as expected.

At that moment I am not sure if we want to try improve this further, or just initial fix (`rbd_options = {NULL, (char*) "rbd", NULL}`) would be enough.

tchaikov commented:
```raise Error('error executing ceph-conf', e, err)``` in the "make check" run

rishabh-d-dave commented:
I don't know why but I don't see stacktrace though I see segfault in gdb as well as dmesg.
[ceph-rbd-fuse-stacktrace-15-Mar-2017.txt](https://github.com/ceph/ceph/files/1815523/ceph-rbd-fuse-stacktrace-15-Mar-2017.txt)

Maybe because I keep getting the message -

```
Missing separate debuginfos, use: dnf debuginfo-install libgcc-7.3.1-2.fc27.x86_64 libstdc++-7.3.1-2.fc27.x86_64 nss-softokn-3.34.0-1.0.fc27.x86_64 nss-softokn-freebl-3.34.0-1.0.fc27.x86_64
```

trociny commented:
did you run "bt" command in gdb after the process terminated (after unmount)?

rishabh-d-dave commented:
> did you run "bt" command in gdb after the process terminated (after unmount)?

Oh, ok. I see the stacktrace. Thanks! I'll try to fix it.

dillaman commented:
@rishabh-d-dave ping

rishabh-d-dave commented:
I've been trying to find out the exact cause. All I understand for now is that message logging happens on a separate thread and running `join` for it leads to `SIGSEV` because the thread handle is [not valid](https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/pthread_join.c;h=6a87a8b329c3e34a57d65e86d45fa97a5fdb2fe2;hb=HEAD#l47).

Copying the stacktrace for reference -

```
Program received signal SIGSEGV, Segmentation fault.
pthread_join (threadid=140660794857216, thread_return=thread_return@entry=0x0) at pthread_join.c:47
47	  if (INVALID_NOT_TERMINATED_TD_P (pd))
(gdb) bt
#0  pthread_join (threadid=140660794857216, thread_return=thread_return@entry=0x0) at pthread_join.c:47
#1  0x00007fee2a52bdec in Thread::join (this=0x55c83b39fc40, prval=prval@entry=0x0) at /home/centos/temp/rbd-fuse-look-for-ceph-conf/ceph/src/common/Thread.cc:164
#2  0x00007fee2a761d43 in ceph::logging::Log::stop (this=<optimized out>) at /home/centos/temp/rbd-fuse-look-for-ceph-conf/ceph/src/log/Log.cc:524
#3  0x00007fee2a571adb in CephContext::~CephContext (this=0x55c83b354250, __in_chrg=<optimized out>) at /home/centos/temp/rbd-fuse-look-for-ceph-conf/ceph/src/common/ceph_context.cc:663
#4  0x00007fee2a571ea4 in CephContext::put (this=0x55c83b354250) at /home/centos/temp/rbd-fuse-look-for-ceph-conf/ceph/src/common/ceph_context.cc:680
#5  0x00007fee33dd0ae3 in operator() (__args#0=<optimized out>, this=<optimized out>) at /opt/rh/devtoolset-7/root/usr/include/c++/7/bits/std_function.h:706
#6  ~unique_ptr (this=0x55c83b3b4070, __in_chrg=<optimized out>) at /opt/rh/devtoolset-7/root/usr/include/c++/7/bits/unique_ptr.h:268
#7  librados::RadosClient::~RadosClient (this=0x55c83b3b4060, __in_chrg=<optimized out>) at /home/centos/temp/rbd-fuse-look-for-ceph-conf/ceph/src/librados/RadosClient.cc:481
#8  0x00007fee33dd0b89 in librados::RadosClient::~RadosClient (this=0x55c83b3b4060, __in_chrg=<optimized out>)
    at /home/centos/temp/rbd-fuse-look-for-ceph-conf/ceph/src/librados/RadosClient.cc:488
#9  0x00007fee33d7d92e in rados_shutdown (cluster=0x55c83b3b4060) at /home/centos/temp/rbd-fuse-look-for-ceph-conf/ceph/src/librados/librados.cc:2915
#10 0x00007fee33af47a2 in fuse_fs_destroy () from /lib64/libfuse.so.2
#11 0x00007fee33af4839 in fuse_lib_destroy () from /lib64/libfuse.so.2
#12 0x00007fee33af6842 in fuse_ll_destroy () from /lib64/libfuse.so.2
#13 0x00007fee33afcdb3 in fuse_session_destroy () from /lib64/libfuse.so.2
#14 0x00007fee33af5b6e in fuse_destroy () from /lib64/libfuse.so.2
#15 0x00007fee33afe130 in fuse_teardown_common () from /lib64/libfuse.so.2
#16 0x00007fee33afe8ae in fuse_main_common () from /lib64/libfuse.so.2
#17 0x000055c839b8a222 in main (argc=2, argv=0x7ffefdf621b8) at /home/centos/temp/rbd-fuse-look-for-ceph-conf/ceph/src/rbd_fuse/rbd-fuse.cc:933
(gdb) q
```

rishabh-d-dave commented:
@trociny @dillaman can you help me out a bit?

dillaman commented:
@rishabh-d-dave I'll take a look today

dillaman commented:
@rishabh-d-dave Try something like this (works for me):

```
diff --git a/src/rbd_fuse/CMakeLists.txt b/src/rbd_fuse/CMakeLists.txt
index 618992b776..e0bd3e2497 100644
--- a/src/rbd_fuse/CMakeLists.txt
+++ b/src/rbd_fuse/CMakeLists.txt
@@ -1,5 +1,5 @@
 add_executable(rbd-fuse
   rbd-fuse.cc)
 target_link_libraries(rbd-fuse
-  ceph-common librbd librados ${FUSE_LIBRARIES})
+  ceph-common librbd librados global ${FUSE_LIBRARIES})
 install(TARGETS rbd-fuse DESTINATION bin)
diff --git a/src/rbd_fuse/rbd-fuse.cc b/src/rbd_fuse/rbd-fuse.cc
index d6d44e051f..4c1aa510cc 100644
--- a/src/rbd_fuse/rbd-fuse.cc
+++ b/src/rbd_fuse/rbd-fuse.cc
@@ -28,9 +28,14 @@
 
 #include "include/compat.h"
 #include "include/rbd/librbd.h"
-
 #include "include/ceph_assert.h"
 
+#include "common/ceph_argparse.h"
+#include "common/ceph_context.h"
+
+#include "global/global_init.h"
+#include "global/global_context.h"
+
 static int gotrados = 0;
 char *pool_name;
 char *mount_image_name;
@@ -781,6 +786,7 @@ static struct fuse_opt rbdfs_opts[] = {
        {"-r %s", offsetof(struct rbd_options, image_name), KEY_RBD_IMAGENAME},
        {"--image=%s", offsetof(struct rbd_options, image_name),
        KEY_RBD_IMAGENAME_LONG},
+       FUSE_OPT_END
 };
 
 static void usage(const char *progname)
@@ -854,18 +860,16 @@ int
 connect_to_cluster(rados_t *pcluster)
 {
        int r;
+       global_init_postfork_start(g_ceph_context);
+       common_init_finish(g_ceph_context);
+       global_init_postfork_finish(g_ceph_context);
 
-       r = rados_create(pcluster, NULL);
+       r = rados_create_with_context(pcluster, g_ceph_context);
        if (r < 0) {
                simple_err("Could not create cluster handle", r);
                return r;
        }
-       rados_conf_parse_env(*pcluster, NULL);
-       r = rados_conf_read_file(*pcluster, rbd_options.ceph_config);
-       if (r < 0) {
-               simple_err("Error reading Ceph config file", r);
-               goto failed_shutdown;
-       }
+
        r = rados_connect(*pcluster);
        if (r < 0) {
                simple_err("Error connecting to cluster", r);
@@ -879,11 +883,42 @@ failed_shutdown:
        return r;
 }
 
-int main(int argc, char *argv[])
+int main(int argc, const char *argv[])
 {
-       struct fuse_args args = FUSE_ARGS_INIT(argc, argv);
+       // librados will filter out -f/-d options from command-line
+       std::map<std::string, bool> filter_args = {
+               {"-f", false},
+               {"-d", false}};
+
+       std::vector<const char*> arg_vector;
+       for (auto idx = 0; idx < argc; ++idx) {
+               auto it = filter_args.find(argv[idx]);
+               if (it != filter_args.end()) {
+                       it->second = true;
+               }
+               arg_vector.push_back(argv[idx]);
+       }
+
+       auto cct = global_init(NULL, arg_vector, CEPH_ENTITY_TYPE_CLIENT,
+                              CODE_ENVIRONMENT_DAEMON,
+                              CINIT_FLAG_UNPRIVILEGED_DAEMON_DEFAULTS);
+       g_ceph_context->_conf.set_val_or_die("pid_file", "");
+       g_ceph_context->_conf.set_val_or_die("daemonize", "true");
+
+       if (global_init_prefork(g_ceph_context) < 0) {
+               fprintf(stderr, "Failed to initialize librados\n");
+               exit(1);
+       }
 
-       if (fuse_opt_parse(&args, &rbd_options, rbdfs_opts, rbdfs_opt_proc) == -1) {
+       for (auto& it : filter_args) {
+               if (it.second) {
+                       arg_vector.push_back(it.first.c_str());
+               }
+       }
+       struct fuse_args args = FUSE_ARGS_INIT((int)arg_vector.size(),
+                                              (char**)&arg_vector.front());
+       if (fuse_opt_parse(&args, &rbd_options, rbdfs_opts,
+                          rbdfs_opt_proc) == -1) {
                exit(1);
        }
 
```

rishabh-d-dave commented:
@dillaman 
Thanks alot. I'll try and let you know.

rishabh-d-dave commented:
@dillaman Once again, thank you for the help. :)

I've spent some with the patch you've provided. It doesn't look for other `ceph.conf` in other standard locations if the one at `/etc/ceph` is absent.  I am trying to implement that but I am not well-acquainted with `src/global` or `src/common`. With respect to the code in the patch above, is the choosing of the right `ceph.conf`, ideally, supposed to happen in `global_init()`?

dillaman commented:
@rishabh-d-dave Hmm -- seemed to work when I ran it locally (since I usually don't have a "/etc/ceph/ceph.conf" file). Basically, ```global_init``` invokes ```global_pre_init``` [1] which invokes ```ceph_argparse_early_args``` [2] which populates the ceph config search list [3] which is then used when loading the configs [4]. If you don't specify a config override on the CLI, it should default [5][6].


[1] https://github.com/ceph/ceph/blob/master/src/global/global_init.cc#L165
[2] https://github.com/ceph/ceph/blob/master/src/global/global_init.cc#L92
[3] https://github.com/ceph/ceph/blob/master/src/common/ceph_argparse.cc#L485
[4] https://github.com/ceph/ceph/blob/master/src/global/global_init.cc#L113
[5] https://github.com/ceph/ceph/blob/master/src/common/config.cc#L346
[6] https://github.com/ceph/ceph/blob/master/src/common/config.cc#L46

rishabh-d-dave commented:
I've updated PR with a minor change (got rid of a goto and corresponding label since it was being jumped to only once) and added your signature to the git commit. I hope you are okay with it.

And I'll address you previous comment shortly.

rishabh-d-dave commented:
Jason Dillaman [wrote](https://github.com/ceph/ceph/pull/20598#issuecomment-432407732):
> @rishabh-d-dave Hmm -- seemed to work when I ran it locally (since I usually don't have a "/etc/ceph/ceph.conf" file). 

Are you sure? I tried again but rbd-fuse seems to be fixated on `/etc/ceph/ceph.conf`. I applied the following patch -

```
diff --git a/src/rbd_fuse/rbd-fuse.cc b/src/rbd_fuse/rbd-fuse.cc
index 35a9944..3169570 100644
--- a/src/rbd_fuse/rbd-fuse.cc
+++ b/src/rbd_fuse/rbd-fuse.cc
@@ -498,9 +498,14 @@ rbdfs_init(struct fuse_conn_info *conn)
        // init cannot fail, so if we fail here, gotrados remains at 0,
        // causing other operations to fail immediately with ENXIO
 
+FILE *fp = fopen("/home/centos/repos/ceph/build/log.debug", "a+");
+fprintf(fp, "before connecting to the cluster - \n"); fflush(fp);
+fprintf(fp, "rbd_options.ceph_conifg = %s\n", rbd_options.ceph_config); fflush(fp);
        ret = connect_to_cluster(&cluster);
        if (ret < 0)
                exit(90);
+fprintf(fp, "after connecting to the cluster - \n"); fflush(fp);
+fprintf(fp, "rbd_options.ceph_conifg = %s\n", rbd_options.ceph_config); fflush(fp);
 
        pool_name = rbd_options.pool_name;
        mount_image_name = rbd_options.image_name;
@@ -896,16 +901,24 @@ int main(int argc, const char *argv[])
                arg_vector.push_back(argv[idx]);
        }
 
+FILE *fp = fopen("/home/centos/repos/ceph/build/log.debug", "a+");
+fprintf(fp, "before global_init - \n"); fflush(fp);
+fprintf(fp, "rbd_options.ceph_conifg = %s\n", rbd_options.ceph_config); fflush(fp);
        auto cct = global_init(NULL, arg_vector, CEPH_ENTITY_TYPE_CLIENT,
                               CODE_ENVIRONMENT_DAEMON,
                               CINIT_FLAG_UNPRIVILEGED_DAEMON_DEFAULTS);
        g_ceph_context->_conf.set_val_or_die("pid_file", "");
        g_ceph_context->_conf.set_val_or_die("daemonize", "true");
 
+fprintf(fp, "before global_init_prefork - \n"); fflush(fp);
+fprintf(fp, "rbd_options.ceph_conifg = %s\n", rbd_options.ceph_config); fflush(fp);
        if (global_init_prefork(g_ceph_context) < 0) {
                fprintf(stderr, "Failed to initialize librados\n");
                exit(1);
        }
+fprintf(fp, "after global_init_prefork - \n"); fflush(fp);
+fprintf(fp, "rbd_options.ceph_conifg = %s\n", rbd_options.ceph_config); fflush(fp);
+fclose(fp);
 
        for (auto& it : filter_args) {
                if (it.second) {
```

ran the following steps after setting up the cluster using `vstart.sh` -
```
./bin/ceph osd pool create rbd 128
./bin/rbd pool init rbd
./bin/rbd create "test-img" --size 100M
sudo mkdir /mnt/rbd_images

sudo ./bin/rbd-fuse /mnt/rbd_images/
```

I see the following text in my debug file -
```
before global_init - 
rbd_options.ceph_conifg = /etc/ceph/ceph.conf
before global_init_prefork - 
rbd_options.ceph_conifg = /etc/ceph/ceph.conf
after global_init_prefork - 
rbd_options.ceph_conifg = /etc/ceph/ceph.conf
before connecting to the cluster - 
rbd_options.ceph_conifg = /etc/ceph/ceph.conf
after connecting to the cluster - 
rbd_options.ceph_conifg = /etc/ceph/ceph.conf
```
And I don't have a `/etc/ceph/ceph.conf`
```
$ ls /etc/ceph
ls: cannot access /etc/ceph: No such file or directory
$ ls ceph.conf
ceph.conf
$
```

dillaman commented:
@rishabh-d-dave It doesn't matter what ```rbd_options.ceph_config``` is set as -- in fact, it should be removed since it isn't used at all anymore.

dillaman commented:
@rishabh-d-dave Note: I just pushed the removal of that variable directly to your branch

rishabh-d-dave commented:
> @rishabh-d-dave It doesn't matter what rbd_options.ceph_config is set as -- in fact, it should be removed since it isn't used at all anymore.

I was trying to see what's going on inside rbd-fuse using gdb; was about to reach the same conclusion.

dillaman commented:
@rishabh-d-dave No need to yank your signature unless you are uncomfortable about it. Thanks for testing all the combinations since rbd-fuse is currently lacks automated testing.

dillaman commented:
retest this please
